packages/contracts/contracts/plugin/PluginRepo.sol

```diff
@@ -118,110 +168,104 @@ contract PluginRepo is
-    /// @return semanticVersion The semantic version number.
-    /// @return pluginSetup The address of the plugin factory associated with the version.
-    /// @return contentURI The external URI pointing to the content of the version.
-    function getVersionByPluginSetup(address _pluginSetup)
-        public
-        view
-        returns (
-            uint16[3] memory semanticVersion,
-            address pluginSetup,
-            bytes memory contentURI
-        )
-    {
-        return getVersionById(versionIndexForPluginSetup[_pluginSetup]);
+    /// @notice get the latest version by plugin setup.
+    /// @param _pluginSetup the plugin setup address
+    /// @return Version latest version that is binded to the _pluginSetup
+    function getLatestVersion(address _pluginSetup) public view returns (Version memory) {
+        return getVersion(latestTagHashForPluginSetup[_pluginSetup]);
     }
 
-    /// @notice Gets the version information associated with a semantic version number.
-    /// @return semanticVersion The semantic version number.
-    /// @return pluginSetup The address of the plugin factory associated with the version.
-    /// @return contentURI The external URI pointing to the content of the version.
-    function getVersionBySemanticVersion(uint16[3] memory _semanticVersion)
+    /// @notice get the version by tag.
+    /// @param _tag the version tag.
+    /// @return the version which belongs to the _tag.
+    function getVersion(Tag calldata _tag)
         public
         view
-        returns (
-            uint16[3] memory semanticVersion,
-            address pluginSetup,
-            bytes memory contentURI
-        )
+        returns (Version memory)
     {
-        return getVersionById(versionIndexForSemantic[semanticVersionHash(_semanticVersion)]);
+        return getVersion(tagHash(_tag));
     }
 
-    /// @notice Gets the version information associated with a version index.
-    /// @return semanticVersion The semantic version number.
-    /// @return pluginSetup The address of the plugin factory associated with the version.
-    /// @return contentURI The external URI pointing to the content of the version.
-    function getVersionById(uint256 _versionIndex)
+    /// @notice get the concrete version.
+    /// @param _tagHash the tag hash.
+    /// @return Version the concrete version by the exact hash.
+    function getVersion(bytes32 _tagHash)
         public
         view
-        returns (
-            uint16[3] memory semanticVersion,
-            address pluginSetup,
-            bytes memory contentURI
-        )
+        returns (Version memory)
     {
-        if (_versionIndex <= 0 || _versionIndex >= nextVersionIndex)
-            revert VersionIndexDoesNotExist({versionIndex: _versionIndex});
-        Version storage version = versions[_versionIndex];
-        return (version.semanticVersion, version.pluginSetup, version.contentURI);
-    }
+        Version memory version = versions[_tagHash];
```

Comment:
use `storage` instead of `memory`

Reply:
Fixed, but not sure If it makes more sense as in the end, copy to memory anyways is needed.